### PR TITLE
Add minimum lowercase rule for password validation.

### PR DESCRIPTION
**Pull request by HLeithner**
Pull Request for Issue #24156 .

### Summary of Changes
Add missing lower case requirement for passwords

### Testing Instructions
Set minimum lowercase count to 0 or higher.
Change password. Use password with more and less lowercase characters

### Expected result
It should be possible to force a minimum of lowercase characters.

### Actual result
Option is missing.

### CODE REVIEW

File: administrator/language/en-GB/en-GB.com_admin.ini [hunk 1]
```diff
@@ -193,6 +193,8 @@ COM_ADMIN_ZLIB_ENABLED="Zlib Enabled"
 ; Messages
 COM_USERS_MSG_NOT_ENOUGH_INTEGERS_N="Password does not have enough digits. At least %s digits are required."
 COM_USERS_MSG_NOT_ENOUGH_INTEGERS_N_1="Password does not have enough digits. At least 1 digit is required."
+COM_USERS_MSG_NOT_ENOUGH_LOWERCASE_LETTERS_N="Password does not have enough lowercase characters. At least %s lower case characters are required."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_admin.ini]: Lower case can be written as either one word or two. At a minimum we should be consistent in only using one variant. Using the gov.uk website as a source we should standardise on one word not two

File: administrator/language/en-GB/en-GB.com_admin.ini [hunk 2]
```diff
@@ -193,6 +193,8 @@ COM_ADMIN_ZLIB_ENABLED="Zlib Enabled"
...
+COM_USERS_MSG_NOT_ENOUGH_LOWERCASE_LETTERS_N_1="Password does not have enough lowercase characters. At least 1 lower case character is required."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_admin.ini]: as above

**Comment by HLeithner** [comment]
I did a copy of the uppercase sentence,.. should be changed too?

**Comment by brianteeman** [comment]
I just realised that too ;)

I don't care if we use one word or two but we should be consistent - especially in the same sentence. Should help the translators as well

File: administrator/language/en-GB/en-GB.com_users.ini [hunk 3]
```diff
@@ -46,6 +46,8 @@ COM_USERS_CONFIG_FIELD_MAILTOADMIN_DESC="If set to Yes then a notification mail
 COM_USERS_CONFIG_FIELD_MAILTOADMIN_LABEL="Send Mail to Administrators"
 COM_USERS_CONFIG_FIELD_MINIMUM_INTEGERS="Minimum Integers"
 COM_USERS_CONFIG_FIELD_MINIMUM_INTEGERS_DESC="Set the minimum number of integers that must be included in a password."
+COM_USERS_CONFIG_FIELD_MINIMUM_LOWERCASE="Minimum Lower Case"
+COM_USERS_CONFIG_FIELD_MINIMUM_LOWERCASE_DESC="Set the minimum number of lower case alphabetical characters required for a password."
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.com_users.ini:50]: Change here? Label too?

**Comment by eric-lamy-agerix** [comment]
Tested with Joomla! Patch Tester and it works fine

**Comment by ghost** [comment]
@agerix please mark your Test at Issue Tracker as "Tested successfully".

**Comment by eric-lamy-agerix** [comment]
I have tested this item :white_check_mark: successfully on ee1d90aa95f64d56d16cf19ace7bc0912129d2e1<br /><br />Hello, I've test with one and with more than one lower case<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24230">issues.joomla.org/tracker/joomla-cms/24230</a>.</sub>

**Comment by eric-lamy-agerix** [comment]
I've just do it Franz, I'm a new tester... :)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24230">issues.joomla.org/tracker/joomla-cms/24230</a>.</sub>

**Comment by HLeithner** [comment]
@quy you are right, I changed all to two words as in the label.

**Comment by dgrammatiko** [comment]
@HLeithner you need to reflect these changes here: https://github.com/joomla/joomla-cms/blob/staging/media/system/js/passwordstrength.js

A fair warning tho, that file is Mootools based so in essence wouldn't be advisable to ask anyone to add any new functionality there. In short better do this in the 4.0 repo...

**Comment by brianteeman** [comment]
This is already available in j4

**Comment by dgrammatiko** [comment]
> This is already available in j4

Well I lost track of the things I did already for J4 😉

Anyways, then at a bare minimum, the last function in the Mootools version needs to be adapted

**Comment by HLeithner** [comment]
@dgrammatiko this file looks like its only show the password strength and doesn't have anything to do with the requirement.

**Comment by dgrammatiko** [comment]
@HLeithner it should read the requirements for calculating the strength, or maybe not...

**Comment by HLeithner** [comment]
The script uses the threshold that is set as parameter I didn't find any calculation based on the requirements parameters...

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on da122f35cb8892196b1ddfdf3ab66a68a5ac397b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24230">issues.joomla.org/tracker/joomla-cms/24230</a>.</sub>

**Comment by jeckodevelopment** [comment]
@agerix can you please submit again your test?

File: libraries/src/Form/Rule/PasswordRule.php [hunk 4]
```diff
@@ -154,6 +157,22 @@ public function test(\SimpleXMLElement $element, $value, $group = null, Registry
...
 		}
 
+		// Minimum number of lower case ASCII characters required
+		if (!empty($minimumLowercase))
+		{
+			$nLowercase = preg_match_all('/[a-z]/', $value, $umatch);
+
+			if ($nLowercase < $minimumLowercase)
+			{
+				\JFactory::getApplication()->enqueueMessage(
```
Comment (alikon) [inline libraries/src/Form/Rule/PasswordRule.php:167]: should be easy to namespace

File: libraries/src/Form/Rule/PasswordRule.php [hunk 5]
```diff
@@ -154,6 +157,22 @@ public function test(\SimpleXMLElement $element, $value, $group = null, Registry
...
+					\JText::plural('COM_USERS_MSG_NOT_ENOUGH_LOWERCASE_LETTERS_N', $minimumLowercase),
```
Comment (alikon) [inline libraries/src/Form/Rule/PasswordRule.php:168]: here too

**Comment by HLeithner** [comment]
If I add the name spaces here I have to change the complete file and I think this would maybe give a merge conflict when george merge it into 4.0

**Comment by brianteeman** [comment]
J4 already has a lowercase rule

**Comment by alikon** [comment]
maybe it's me but the 4.0 https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/Form/Rule/PasswordRule.php didn't have lowercase yet....

@HLeithner not an expert on "easy merging"  matter  but looking the 2 file side by side ....

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on e3869297be333bb69e6eb569329ad746a143e962<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24230">issues.joomla.org/tracker/joomla-cms/24230</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on e3869297be333bb69e6eb569329ad746a143e962<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24230">issues.joomla.org/tracker/joomla-cms/24230</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24230">issues.joomla.org/tracker/joomla-cms/24230</a>.</sub>

**Pull request merged**